core/src/main/java/com/linecorp/armeria/common/stream/HttpDeframer.java

```diff
@@ -0,0 +1,268 @@
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A skeletal {@link Processor} implementation that decodes a stream of {@link HttpObject}s to N objects.
+ */
+@UnstableApi
+public abstract class HttpDeframer<T> extends DefaultStreamMessage<T> implements Processor<HttpObject, T> {
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<HttpDeframer, Subscription> upstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(HttpDeframer.class, Subscription.class, "upstream");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<HttpDeframer> initializedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(HttpDeframer.class, "initialized");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<HttpDeframer> subscribedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(HttpDeframer.class, "subscribed");
+
+    private final ByteBufDeframerInput input;
+    private final EventExecutor eventLoop;
+    private final ArrayDeque<T> outputQueue = new ArrayDeque<>();
+
+    private boolean sawLeadingHeaders;
+
+    @Nullable
+    private volatile Subscription upstream;
+    private volatile boolean cancelled;
+    private volatile int initialized;
+    private volatile int subscribed;
+
+    /**
+     * Returns a new {@link HttpDeframer} with the specified {@link EventExecutor} and
+     * {@link ByteBufAllocator}.
+     */
+    protected HttpDeframer(EventExecutor eventLoop, ByteBufAllocator alloc) {
+        requireNonNull(eventLoop, "eventLoop");
+        requireNonNull(alloc, "alloc");
+        input = new ByteBufDeframerInput(alloc);
+        this.eventLoop = eventLoop;
+    }
+
+    /**
+     * Converts the specified {@link HttpData} to {@link ByteBuf}.
+     */
+    protected ByteBuf convertToByteBuf(HttpData data) {
+        return data.byteBuf();
+    }
+
+    /**
+     * Decodes a stream of {@link HttpData}s to N objects.
+     * This method will be called whenever an {@link HttpData} is signaled from {@link Publisher}.
+     */
+    protected abstract void process(HttpDeframerInput in, HttpDeframerOutput<T> out);
+
+    private void process(HttpObject data) {
+        if (data instanceof HttpHeaders) {
+            final HttpHeaders headers = (HttpHeaders) data;
+
+            if (headers instanceof ResponseHeaders) {
+                final ResponseHeaders responseHeaders = (ResponseHeaders) headers;
+                if (responseHeaders.status().isInformational()) {
+                    processInformationalHeaders(responseHeaders, outputQueue::addLast);
+                    return;
+                }
+            }
+
+            if (!sawLeadingHeaders) {
+                sawLeadingHeaders = true;
+                processHeaders((HttpHeaders) data, outputQueue::addLast);
```

Comment:
We can move this into the above if statement and change the method parameter to use `ResponseHeaders`.
`processHeaders(ResponseHeaders in, HttpDeframerOutput<T> out)`

Reply:
HttpDeframer is used for both an HTTP request and response so `processHeaders` should take HttpHeaders and handle it. 😉